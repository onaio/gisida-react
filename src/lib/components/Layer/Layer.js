import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Actions, prepareLayer, lngLat } from 'gisida';


const mapStateToProps = (state, ownProps) => {
  const { APP, LOC } = state;
  const MAP = state[ownProps.mapId]
  if (!!!MAP) {
    throw "MAP not found"
  }
  return {
    APP,
    LOC,
    timeSeriesObj: MAP.timeseries[MAP.visibleLayerId],
    timeseries: MAP.timeseries,
    layers: MAP.layers
  }
}

export class Layer extends Component {

  onLayerToggle = layer => {
    // dispatch toggle layer action
    const { mapId, APP, LOC } = this.props;
    if (!mapId) {
      return null;
    }
    this.props.dispatch(Actions.toggleLayer(mapId, layer.id));
    const {center, zoom } = lngLat(LOC, APP);
    if (layer.zoomOnToggle && layer.visible) {
        window.maps.forEach((e) => {
          e.easeTo({
            center,
            zoom,
          })
        });  
      } 
    // Prepare layer if layer had not been loaded
    if (!layer.loaded && !layer.isLoading) {
      prepareLayer(mapId, layer, this.props.dispatch);
    }
  }

  render() {
    const layer = this.props.layer;
    const mapId = this.props.mapId;
    if (!mapId) {
      return null;
    }
    return (
      <li className="layer" title={layer['menu-credit'] ? `Credit: ${layer['menu-credit']}` : ''}>
        <input
          id={`${layer.id}-${mapId}`}
          type="checkbox"
          data-layer={layer.id}
          onChange={e => this.onLayerToggle(layer)}
          checked={!!layer.visible}
        />
        <label htmlFor={`${layer.id}-${mapId}`}>{layer.label}</label>
      </li>
    );
  }
}

Layer.propTypes = {
  layer: PropTypes.objectOf(PropTypes.any).isRequired,
  mapId: PropTypes.string,
};

export default connect(mapStateToProps)(Layer);
