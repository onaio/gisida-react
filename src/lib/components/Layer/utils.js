import { QUERY_PARAM_LAYERS } from '../../constants';
import { pushSearchParamsToURL, getURLSearchParams } from '../../utils';

/**
 * Push layer to URL which can be used for sharing
 * @param {*} layer
 */
export const pushLayerToURL = (layer, mapId) => {
  const layerId = layer.id.replace('.json', '');
  const queryParamLayers = `${mapId}-${QUERY_PARAM_LAYERS}`;
  const urlSearchParams = getURLSearchParams();
  /**
   * Check for visibility. If false it means layer has been selected push to layer id to url
   * else if visible it means layer has been checked off pop layer from url
   */
  if (layer && layer.visible === false) {
    if (urlSearchParams.has(queryParamLayers)) {
      /**
       * If query param is in URL, append the layer id
       */
      urlSearchParams.append(queryParamLayers, layerId);
    } else {
      /**
       * If query param is not in URL, set the value as the first
       * value of the query param
       */
      urlSearchParams.set(queryParamLayers, layerId);
    }
  } else if (layer && layer.visible === true) {
    if (urlSearchParams.has(queryParamLayers)) {
      /**
       * We filter out the layer and reset the query param
       * with the remainder
       */
      const remainingLayers = urlSearchParams.getAll(queryParamLayers).filter(l => l !== layerId);
      urlSearchParams.delete(queryParamLayers);
      remainingLayers.forEach(val => {
        urlSearchParams.append(queryParamLayers, val);
      });
    }
  }

  pushSearchParamsToURL(urlSearchParams);
};
