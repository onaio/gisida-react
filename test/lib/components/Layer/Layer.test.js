import React from 'react';
import { Layer }  from '../../../../src/lib/components/Layer/Layer'
import layerObj from '../../../fixtures/sample-layer.json';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';


layerObj.id = 'sample-layer';
const mapId = 'map-1';
const componentWrapper = shallow(
  <Layer
    layer={layerObj}
    mapId={mapId}
  />
);

describe('Layer', () => {
  it('component renderes correctly', () => {
    const json = toJson(componentWrapper)
    expect(json).toMatchSnapshot();
  });

  it('onLayerToggle is called when layer is checked', () => {
    const event = {
      target:
        { checked: true }
    }
    componentWrapper.instance().onLayerToggle = jest.fn();
    componentWrapper.update();
    componentWrapper.find(`#${layerObj.id}-${mapId}`).simulate('change', event);
    expect(componentWrapper.instance().onLayerToggle).toBeCalledWith(layerObj);
  })
});
