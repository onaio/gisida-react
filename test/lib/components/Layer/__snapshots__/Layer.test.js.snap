// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layer component renderes correctly 1`] = `
<li
  className="layer"
  title=""
>
  <input
    checked={true}
    data-layer="sample-layer"
    id="sample-layer-map-1"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    htmlFor="sample-layer-map-1"
  >
    Boundaries
  </label>
</li>
`;
